modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/IgniteAbstractTxSuspendResumeTest.java

```diff
@@ -0,0 +1,771 @@
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.suspend();
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.close();
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.commit();
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.commitAsync().get(FUT_TIMEOUT);
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.rollback();
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.rollbackAsync().get(FUT_TIMEOUT);
+            }
+        },
+        new CI1Exc<Transaction>() {
+            @Override public void applyx(Transaction tx) throws Exception {
+                tx.setRollbackOnly();
+            }
+        }
+    );
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        int idx = getTestIgniteInstanceIndex(igniteInstanceName);
+
+        boolean client = idx >= SERVER_CNT && idx < GRID_CNT;
+
+        cfg.setClientMode(client);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        startGridsMultiThreaded(gridCount());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        Ignite client = ignite(gridCount() - 1);
+
+        assertTrue(client.cluster().localNode().isClient());
+
+        for (CacheConfiguration<Integer, Integer> ccfg : cacheConfigurations()) {
+            grid(0).createCache(ccfg);
+
+            if (ccfg.getCacheMode() != LOCAL && !FORCE_MVCC)
+                client.createNearCache(ccfg.getName(), new NearCacheConfiguration<>());
+        }
+
+        awaitPartitionMapExchange();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        for (CacheConfiguration ccfg : cacheConfigurations())
+            ignite(0).destroyCache(ccfg.getName());
+
+        super.afterTest();
+    }
+
+    /**
+     * @return Number of server nodes.
+     */
+    protected int gridCount() {
+        return GRID_CNT;
+    }
+
+    /**
+     * @return Transaction concurrency.
+     */
+    protected abstract TransactionConcurrency transactionConcurrency();
```

Comment:
Minor.
"@Parametrized" can be used instead.

Reply:
AFAIK "@Paramentrized" can be used if we need to run some set of tests with different parameters, 
but here we have different sets of tests for OPTIMISTIC and PESSIMISTIC transactions.

Reply:
Tests that not supposed to use such concurrency level can be skipped (I see few counts of them)
Moreover, I see that almost all tests use the following pattern:
peek node, peek cache configuration, peek tx isolation level -> doTest.
Can we somehow replace it with `@Parametrized`?
I see that only one test doesn't match this pattern `testSuspendTxAndResumeAfterTopologyChange`. It can be separated, while others can user test parameters. This will avoid copypaste and every test will be plain.

Reply:
`IgniteAbstractTxSuspendResumeTest` is not a new test. The test was renamed from `IgniteOptimisticTxSuspendResumeTest`. It already has some logic which we can accidentally break by deep refactoring, it also has some history on TC, which can be lost.
I see no practical reason to totally rewrite the test only to use some JUnit feature.

Reply:
This is not "some JUnit feature", this is the right way to write tests without boilerplate.
But I agree that the existing test with history is a restrictive factor of refactoring.
Anyway, I recommend writing tests with `@Parametrized` pattern in future tests if applicable.

Reply:
Ok, I will rewrite `GridJtaTransactionManagerSelfTest` with '@Parametrized', but this test I prefer to left as is.